File: mobile/app/src/main/java/com/ticketbus/mobile/offline/SyncWorker.java
package com.ticketbus.mobile.offline;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.ticketbus.mobile.data.local.AppDatabase;
import com.ticketbus.mobile.data.local.entity.BlacklistEntryLocal;
import com.ticketbus.mobile.data.local.entity.PendingValidation;
import com.ticketbus.mobile.data.remote.ApiService;
import com.ticketbus.mobile.data.remote.RetrofitClient;
import com.ticketbus.mobile.data.remote.dto.BlacklistDto;
import com.ticketbus.mobile.data.remote.dto.PendingValidationDto;
import com.ticketbus.mobile.data.remote.dto.PublicKeyDto;
import com.ticketbus.mobile.data.remote.dto.SyncResponse;
import com.ticketbus.mobile.data.remote.dto.SyncUploadRequest;
import com.ticketbus.mobile.util.PreferenceManager;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

import retrofit2.Response;

public class SyncWorker extends Worker {
    private static final String TAG = "SyncWorker";

    public SyncWorker(@NonNull Context context, @NonNull WorkerParameters params) {
        super(context, params);
    }

    @NonNull
    @Override
    public Result doWork() {
        Context ctx = getApplicationContext();
        PreferenceManager prefs = new PreferenceManager(ctx);
        AppDatabase db = AppDatabase.getInstance(ctx);
        ApiService api = RetrofitClient.create(prefs.getAuthToken());

        // Upload pending validations
        List<PendingValidation> pending = db.validationEventDao().getUnsynced();
        if (!pending.isEmpty()) {
            List<PendingValidationDto> dtos = new ArrayList<>();
            for (PendingValidation pv : pending) {
                dtos.add(new PendingValidationDto(
                    "", pv.deviceId, pv.location, pv.latitude, pv.longitude,
                    Instant.ofEpochMilli(pv.validationTime).toString(), pv.result
                ));
            }
            SyncUploadRequest req = new SyncUploadRequest(
                prefs.getUsername() != null ? prefs.getUsername() : "unknown",
                dtos,
                prefs.getLastSync()
            );
            try {
                Response<SyncResponse> resp = api.uploadSync(req).execute();
                if (resp.isSuccessful()) {
                    for (PendingValidation pv : pending) {
                        db.validationEventDao().markSynced(pv.id);
                    }
                }
            } catch (Exception e) {
                Log.w(TAG, "Upload failed", e);
            }
        }

        // Fetch blacklist updates
        try {
            Response<List<BlacklistDto>> resp = api.getBlacklist(prefs.getLastSync()).execute();
            if (resp.isSuccessful() && resp.body() != null) {
                List<BlacklistEntryLocal> entries = new ArrayList<>();
                for (BlacklistDto dto : resp.body()) {
                    entries.add(new BlacklistEntryLocal(dto.ticketNumber, dto.reason));
                }
                if (!entries.isEmpty()) {
                    db.blacklistDao().insertAll(entries);
                }
            }
        } catch (Exception e) {
            Log.w(TAG, "Blacklist fetch failed", e);
        }

        // Fetch public key
        try {
            Response<PublicKeyDto> resp = api.getPublicKey().execute();
            if (resp.isSuccessful() && resp.body() != null) {
                prefs.setPublicKey(resp.body().key);
            }
        } catch (Exception e) {
            Log.w(TAG, "Public key fetch failed", e);
        }

        prefs.setLastSync(Instant.now().toString());
        return Result.success();
    }
}

File: mobile/app/src/main/java/com/ticketbus/mobile/offline/SyncWorker.kt
package com.ticketbus.mobile.offline

import android.content.Context
import androidx.work.*
import com.ticketbus.mobile.data.local.AppDatabase
import com.ticketbus.mobile.data.local.entity.BlacklistEntryLocal
import com.ticketbus.mobile.data.remote.RetrofitClient
import com.ticketbus.mobile.data.remote.dto.*
import com.ticketbus.mobile.util.PreferenceManager
import java.time.Instant

class SyncWorker(context: Context, params: WorkerParameters) : CoroutineWorker(context, params) {
    override suspend fun doWork(): Result {
        val prefs = PreferenceManager(applicationContext)
        val db = AppDatabase.getInstance(applicationContext)
        val api = RetrofitClient.create(prefs.authToken)

        // Upload pending validations
        val pending = db.validationEventDao().getUnsynced()
        if (pending.isNotEmpty()) {
            val dtos = pending.map {
                PendingValidationDto(
                    ticketNumber = "", deviceId = it.deviceId, location = it.location,
                    latitude = it.latitude, longitude = it.longitude,
                    validationTime = Instant.ofEpochMilli(it.validationTime).toString(),
                    result = it.result
                )
            }
            val req = SyncUploadRequest(prefs.username ?: "unknown", dtos, prefs.lastSync)
            try {
                val resp = api.uploadSync(req)
                if (resp.isSuccessful) pending.forEach { db.validationEventDao().markSynced(it.id) }
            } catch (_: Exception) {}
        }

        // Fetch blacklist updates
        try {
            val resp = api.getBlacklist(prefs.lastSync)
            if (resp.isSuccessful) {
                val entries = resp.body()?.map { BlacklistEntryLocal(it.ticketNumber, it.reason) } ?: emptyList()
                if (entries.isNotEmpty()) db.blacklistDao().insertAll(entries)
            }
        } catch (_: Exception) {}

        // Fetch public key
        try {
            val resp = api.getPublicKey()
            if (resp.isSuccessful) prefs.publicKey = resp.body()?.key
        } catch (_: Exception) {}

        prefs.lastSync = Instant.now().toString()
        return Result.success()
    }
}
